From e5f9beaa005962c890a5aa535f8efb8ed4f3d7d6 Mon Sep 17 00:00:00 2001
From: Ratchanan Srirattanamet <ratchanan@ubports.com>
Date: Fri, 1 Aug 2025 21:02:09 +0700
Subject: [PATCH 2/2] modules/MirSurface: try to let Mir forceClose dead
 surface close timeout

This is the last-ditch attempt on this weird behavior. This really a
workaround; we still have to figure out what actually happen.

Signed-off-by: Mike Gabriel <mike.gabriel@das-netzwerkteam.de>
---
 src/modules/QtMir/Application/mirsurface.cpp | 13 +++++++++++++
 1 file changed, 13 insertions(+)

--- a/src/modules/QtMir/Application/mirsurface.cpp
+++ b/src/modules/QtMir/Application/mirsurface.cpp
@@ -1203,11 +1203,10 @@
                 // to kill the app.
                 // ref https://github.com/ubports/ubuntu-touch/issues/1417
                 WARNING_MSG << "(), app with ID " << app->appId() << " has " <<
-                    "ignored request to close a window. Tearing down window. " <<
-                    "This could be a bug in the application.";
-                //app->terminate();
-                teardown();
-            } // TODO: Investigate phone/desktop split
+                    "ignored request to close a window. Terminating the " <<
+                    "application. This could be a bug in the application.";
+                app->terminate();
+            }
         } else {
             // Weird zombie surface. Removing it may still cause problems later
             // since this should almost never occur.
@@ -1216,11 +1215,30 @@
             m_controller->forceClose(m_window);
         }
     } else {
-        WARNING_MSG << "Surface not live anymore, tearing down.";
-        teardown();
+        // In case this is a weird combie process, just try to remove the application it was related to
+        if (m_session && m_session->application()) {
+            Application *app = static_cast<Application*>(m_session->application());
+
+            WARNING_MSG << "(), app with ID " << app->appId() << " has " <<
+                    "ignored request to close a window. Maybe it is dead already?" <<
+                    "Terminating the application anyway.";
+            app->terminate();
+        } else {
+            WARNING_MSG << "() object is still alive despite being not live "
+                    "and has no session. Something hold us up? Will now try to "
+                    "ask Mir to force close this surface. Expect strange "
+                    "behavior.";
+
+            try {
+                m_controller->forceClose(m_window);
+            } catch (const std::out_of_range&) {
+                WARNING_MSG << "() and Mir doesn't even have this surface "
+                        "anymore. What on earth is happenning???";
+                // At this point there's nothing else we can do...
+            }
+        }
     }
 }
-
 void MirSurface::setCloseTimer(AbstractTimer *timer)
 {
     bool timerWasRunning = false;
